### DISCUSSION TITLE
`beforeEach` across all tests?

### DISCUSSION THREAD

**Discussion by SevenOutman** [avajs/ava#2795] [Q&A]
Hi, I wonder how I can setup some common `beforeEach` and `afterEach` hooks for all my test files in one place?

**Comment by novemberborn** [comment]
Could you elaborate on your use cases? There is no explicit mechanism for this.

As you've probably realised, each test file is executed in a child process (and with the upcoming AVA 4, in a worker thread). Perhaps you can use the `require` option and include a helper file in each child process (or worker thread)?

We do have an (experimental) way for test files to communicate with each other, which helps with certain use cases. It would be great to expand this so you can load some code into the AVA lifecycle outside of a test file.

**Reply by SevenOutman** [reply]
I need to do some common setup work for all my test cases (which are located in multiple files), e.g. creating a sandbox folder before each test case is run and removing it after test case is finished. 

Currently I’m including a helper module from which a `setup` function is exported. The inconvenience is that I have to exclude this helper file from test file patterns to avoid an unexpected “no test found in xxx file” warning. It’s not a real problem but I wonder if there’re approaches more elegant.

**Reply by novemberborn** [reply]
No that sounds about right. I'd use the `require` option so you don't have to load it manually in each test file (assuming it applies to each test file).

AVA does have the concept of helper files. If you prefix the helper file with an underscore it's  recognized as such, and AVA should recognize "helper" directories too.

**Reply by SevenOutman** [reply]
> I'd use the `require` option so you don't have to load it manually in each test file

I tried using the `require` option and removing manual imports, but the result was "No tests found" being thrown from every test file. FYI I'm writing tests in ESM and setting all of my test cases to run serially with `test.serial` and `--serial` flag.

> If you prefix the helper file with an underscore it's recognized as such, and AVA should recognize "helper" directories too.

The naming convention works, but is it introduced somewhere in the docs?

**Reply by novemberborn** [reply]
> The naming convention works, but is it introduced somewhere in the docs?

Technically yes, but it's probably not easy to find. The file selection behavior is described here: https://github.com/avajs/ava/blob/main/docs/05-command-line.md



> > I'd use the `require` option so you don't have to load it manually in each test file
> 
> I tried using the `require` option and removing manual imports, but the result was "No tests found" being thrown from every test file.

Hmm, it could be that there's an error (that's not being reported properly) stopping AVA from loading your test file. And then you get the misleading "no tests found" error. I'd have to see a reproduction to dig into that.

**Reply by SevenOutman** [reply]
> Hmm, it could be that there's an error (that's not being reported properly) stopping AVA from loading your test file. And then you get the misleading "no tests found" error. I'd have to see a reproduction to dig into that.

Yes that sounds like what’s happening in my case because I tried adding some `console.debug`s to my test files (outside of test cases) but none of them gets executed.
